### ISSUE TITLE
Flatten solver hierarchy for partitioned-heat-conduction

### ISSUE THREAD

**Issue by uekerman** [precice/tutorials#137]
As part of the [tutorials restructuring](https://github.com/orgs/precice/projects/5) we want to flatten the second (solver) hierarchy. For the `partitioned-heat-conduction` case this is non-trivial.

We currently have: https://github.com/precice/tutorials/tree/restructure/partitioned-heat-conduction
```
- fenics-fenics 
```
We want instead:
```
- images/
- dirichlet-fenics
- neumann-fenics
- dirichlet-nutils
- neumann-nutils
- precice-config.xml
- clean.sh
- README.md
```

The Nutils case: [coupled.zip](https://github.com/precice/tutorials/files/5724131/coupled.zip) (a deprecated version of the Python bindings is still used)

FEniCS and Nutils need to use the same parameters. Dirichlet/Neumann needs to be fixed to be the left or right part etc.

For further conventions please copy from the `turek-hron-fsi3` structure: https://github.com/precice/tutorials/tree/restructure/turek-hron-fsi3

**Comment by BenjaminRueth** [comment]
One important thing to keep in mind: Currently `partitioned-heat-conduction` uses a [vector-valued flux](https://github.com/precice/tutorials/blob/restructure/partitioned-heat-conduction/fenics-fenics/precice-config.xml#L12) and `flow-over-heated-plate` has a [scalar-valued flux](https://github.com/precice/tutorials/blob/1f814630d06a3ada5a0a11689e6a81fd360cd20c/flow-over-heated-plate/buoyantPimpleFoam-fenics/precice-config.xml#L12) (only x component). In both cases the same method is used by the solver to compute the flux (see [here](https://github.com/precice/tutorials/blob/1f814630d06a3ada5a0a11689e6a81fd360cd20c/flow-over-heated-plate/buoyantPimpleFoam-fenics/Solid/heat.py#L63-L77) and [here](https://github.com/precice/tutorials/blob/1f814630d06a3ada5a0a11689e6a81fd360cd20c/partitioned-heat-conduction/fenics-fenics/heat.py#L40-L53)).

Why do we use a vector valued flux? Because this allows us to also support arbitrary geometries (and get the exact solution) for `partitioned-heat-conduction`. See https://github.com/precice/tutorials/pull/46/commits/23e450bb49b35ad28a4d53a19493c9090e4d709d and related PR.

Another thing I don't like so much is that we now get `dirichlet-fenics` and `neumann-fenics` (both `.py`?). This looks to me like a lot of code duplication. Currently we can solve all cases (complex geometry, Neumann/Dirichlet vs. Left/Right) with one single `heat.py` - ok: containing a lot of `if`s that do not help with readability (we can discuss this in a separate issue, see below).

### Different Issue? *simplify `heat.py` and make `partitioned-heat-conduction` and `flow-over-heated-plate` consistent*

Boiling `heat.py` down to the basic functionality and moving the advanced setup somewhere else might be a good idea, if we want to provide a simple tutorial that is short and readable. This would also help us to avoid the vector vs. scalar valued problem, since for `partitioned-heat-conduction` (with simple geometry) we use the x-component of the flux. For the `flow-over-heated-plate` we use the y-component of the flux. We can also just use the x-axis as coupling interface for `partitioned-heat-conduction` then both cases will use the y-component of the flux and the two solvers are really the same.

**Comment by uekerman** [comment]
Consistency between `flow-over-heated-plate` and `partitioned-heat-conduction` would be a nice addition, yes, but not the main concern here.

The Nutils script use scalar fluxes, so I would recommend to use scalar fluxes for FEniCS as well. Simply the flux in normal direction.

> Another thing I don't like so much is that we now get dirichlet-fenics and neumann-fenics (both .py?). 

Both should be subfolders. Compare the FSI3 case mentioned above.

> This looks to me like a lot of code duplication. Currently we can solve all cases (complex geometry, Neumann/Dirichlet vs. Left/Right) with one single heat.py - ok: containing a lot of ifs that do not help with readability (we can discuss this in a separate issue, see below).

Yes, this is exactly the trade-off we need to decide. Let's do this here, it's the main ingredient for the flattening.

**Comment by BenjaminRueth** [comment]
> Yes, this is exactly the trade-off we need to decide. Let's do this here, it's the main ingredient for the flattening.

For the `partitioned-heat-conduction` this boils down to the question what we want to show/do with a tutorial.

1) one single case that a user can run out of the box. Illustration of one feature/physics.
2) a flexible case that can also be used to test certain functionality and do research.

Currently `partitioned-heat-conduction` clearly belongs to the second class. Therefore, the file `heat.py` is quite verbose (many, many `if`s). But it can also be used for a wide range of experiments (subcycling, higher-order time stepping, complex geometry, exchange Dirichlet & Neumann solvers...).

If we prefer tutorials belonging to class 1), I can prepare a draft with the minimal functionality (y-axis as coupling interface, hard-coded rhs, no support for subcycling, position of Dirichlet and Neumann solver fixed; parallel functionality, since this comes with no overhead). This will make the code much more readable. However, then we also need a plan what we want to do with all the nice features mentioned above. Throwing them away would be a waste, since we still use the advanced cases for testing and research. Putting the advanced `heat.py` into an independent repository might be a good solution (similar to the elastictube?).

**Comment by BenjaminRodenberg** [comment]
Some conclusions from personal discussion:

* We will have `partitioned-heat-conduction` only containing the simple geometry case without subcycling. As simple as it can get. This will be part of #125.
* We will have another `partitioned-heat-conduction-complex` (name to be discussed), where we only support the complex geometry case. This is a bit more complicated due to more complex geometry and some additional necessary tweaking. But still: boiled down to the basics. This case will be added after #125 is merged.

I can take care of this.

**Comment by MakisH** [comment]
I understand that this is now solved in develop, please reopen if not.
